docs/rfc/rfc-002-zero-copy-encoding.md

```diff
@@ -0,0 +1,390 @@
+message fields as pointers (see below) if the desire is to save memory when they are rarely used needed.)
+
+#### Oneof’s
+
+`oneof`s are encoded as a combination of a `uint8` discriminant field and memory that is as large as the largest member
+field. `oneof` field numbers *MUST* be between `1` and `255`.
+
+```protobuf
+message Foo {
+  oneof sum {
+    bool x = 1;
+    int32 y = 2;
+  }
+}
+```
+
+A discriminant of `0` indicates that the field is not set.
+
+#### Pointer Types: Bytes and Strings and Repeated fields
+
+A pointer is an 16-bit unsigned integer that points to an offset in the current memory buffer or to another memory
+buffer.  If the bit mask `0xFF00` on the is unset, then the pointer points to an offset in the main 64kb memory buffer.
+If that bit mask is set, then the pointer points to a large `string` or `bytes` buffer.  Up to 256 such buffers
+can be referenced in a single `Root`. The pointer `0` indicates that a field is not defined.
+
+`bytes`, `string`s and repeated fields are encoded as pointers to a memory location that is prefixed with the
+length of the `bytes`, `string` or repeated field value. If the referenced memory location is in the main 64kb memory
+buffer, then this length prefix will be a 16-bit unsigned integer. If the referenced memory location is a large
+`string` or `bytes` buffer, then this length prefix will be a 32-bit unsigned integer.
+
+#### `Any`s
+
+`Any`s are encoded as a pointer to the type URL string and a pointer to the start of the message
+specified by the type URL.
+
+#### Maps
+
+Maps are not supported.
+
+#### Extended Encoding Options
+
+We may choose to allow customizing the encoding of fields so that they take up less space.
+
+For example, we could allow 8-bit or 16-bit integers:
+`int32 x = 1 [(cosmos_proto.int16) = true]` would indicate that the field only needs 2 bytes
+
+Or we could allow `string`, `bytes` or `repeated` fields to have a fixed size rather than being encoding as
+pointers to a variable-length value:
+`string y = 2 [(cosmos_proto.fixed_size) = 3]` could indicate that this is a fixed width 3 byte string
+
+If we choose to enable these encoding options, changing these options would be a breaking change that needs to be
+prevented by the breaking change detector.
+
+### Generated Code
+
+We will describe the generated Go and Rust code using this example protobuf file:
+
+```protobuf
+message Foo {
+  int32 x = 1;
+  optional uint32 y = 2;
+  string z = 3;
+  Bar bar = 4;
+  repeated Bar bars = 5;
+}
+
+
+message Bar {
+  ABC abc = 1;
+  Baz baz = 2;
+  repeated uint32 xs = 3;
+}
+
+message Baz {
+  oneof sum {
+    uint32 x = 1;
+    string y = 2;
+  }
+}
+
+enum ABC {
+  A = 0;
+  B = 1;
+  C = 2;
+  D = 3;
+}
+```
+
+#### Go
+
+In golang, the generated code would not expose any exported struct fields, but rather getters and setters as an
+interface
+or struct methods, ex:
```

Comment:
do we anticipate multiple impl of a given message interface? if not struct methods seem better.